src/back/spv/writer.rs

```diff
@@ -172,52 +207,73 @@ impl Writer {
         class: spirv::StorageClass,
     ) -> Word {
         let ty = &arena[handle];
-        let type_id = self.get_type_id(arena, handle);
-        match ty.inner {
-            crate::TypeInner::Pointer { .. } => type_id,
-            _ => {
+        match self.lookup_pointer.get_id(&LookupType::Local(ty)) {
+            Some(id) => *id,
+            None => {
                 let pointer_id = self.generate_id();
+                let type_id = self.get_type_id(arena, LookupType::Handle(handle));
+
                 let instruction = self.instruction_type_pointer(pointer_id, class, type_id);
                 instruction.to_words(&mut self.logical_layout.declarations);
 
-                /* TODO
-                    Not able to lookup Pointer, because there is no Handle in the IR for it.
-                    Idea would be to not have any handles at all in the lookups, so we aren't bound
-                    to the IR. We can then insert, like here runtime values to the lookups
-                */
-                // self.lookup_type.insert(pointer_id, global_variable.ty);
+                self.lookup_pointer.insert(pointer_id, LookupType::Local(ty));
                 pointer_id
             }
         }
     }
 
-    fn find_scalar_handle(
-        &self,
+    fn get_scalar_id(
+        &mut self,
         arena: &crate::Arena<crate::Type>,
         kind: crate::ScalarKind,
         width: u8,
-    ) -> crate::Handle<crate::Type> {
-        let mut scalar_handle = None;
-        for (handle, ty) in arena.iter() {
-            match ty.inner {
+    ) -> Word {
+        let (mut ty, mut handle) = (None, None);
+        for (lookup_handle, lookup_ty) in arena.iter() {
+            match lookup_ty.inner {
                 crate::TypeInner::Scalar {
                     kind: lookup_kind,
                     width: lookup_width,
                 } => {
                     if kind == lookup_kind && width == lookup_width {
-                        scalar_handle = Some(handle);
+                        ty = Some(lookup_ty);
+                        handle = Some(lookup_handle);
                         break;
                     }
                 }
                 _ => continue,
             }
         }
-        scalar_handle.unwrap()
+
+        // Scalar is generated in the IR
+        if ty.is_some() {
+            match self.lookup_type.get_id(&LookupType::Handle(handle.unwrap())) {
+                Some(id) => *id,
+                None => {
+                    let (instruction, id) = self.parse_type_declaration(arena, LookupType::Handle(handle.unwrap()));
+                    instruction.to_words(&mut self.logical_layout.declarations);
+                    id
+                }
+            }
+        } else {
+            let new_ty: &'a crate::Type = &crate::Type {
+                name: None,
+                inner: crate::TypeInner::Scalar {
+                    width,
+                    kind
+                },
+            };
```

Comment:
I try to create a new type here, but I am not sure why it does not work. Getting this lifetime error:
![image](https://user-images.githubusercontent.com/6235760/89328121-2ce29100-d68d-11ea-9f85-49cc880f471c.png)

Reply:
You are constructing the type on the stack, and then getting a reference to it. This doesn't fulfill "`a".

I think your current `LookupType::Local` variant is dysfunctional: the only thing in "'a" that owns the types is the module's arenat, and if you have the type there, you might as well use `LookupType::Handle`.